Filtering by multiple custom properties

macobo opened the pull request:
### Changes

This PR adds support to allow:
- filtering by multiple custom properties (both frontend and backend)
- (Under behavioral properties view) filtering by other properties when custom property filter applied

Follow-up to https://github.com/plausible/analytics/pull/3708

Constraints:
- The same property cannot be filtered on multiple times (as that would require changing the request/query data schema)

Other notes:
- I'm not super happy with the UI/UX. Let me know if you have suggestions!
- I have yet to update the docs - looking to pair with someone for that?

<details><summary>Click here to see some screenshots!</summary>

![image](https://github.com/plausible/analytics/assets/148820/78d34483-a361-445c-9bb9-cf4baf998e17)

![image](https://github.com/plausible/analytics/assets/148820/2264530c-9f1b-4682-bafb-e0a38b4c3959)

</details>

### Tests
- [x] Automated tests have been added
- [ ] This PR does not require tests

### Changelog
- [x] Entry has been added to changelog
- [ ] This PR does not make a user-facing change

### Documentation
- [ ] [Docs](https://github.com/plausible/docs) have been updated
- [ ] This change does not need a documentation update

RobertJoonas commented on assets/js/dashboard/components/combobox.js:
```diff
@@ -92,7 +92,7 @@ export default function PlausibleCombobox(props) {
 
   function isDisabled(option) {
-    return props.values.some((val) => val.value === option.value)
+    return props.values.some((val) => val.value === option.value) || (props.disabledOptions || []).some((val) => val?.value === option.value)
```
This function is pretty difficult to read. What do you think about extracting some boolean variables here to make it clearer?

Also, the `isDisabled` name is currently used in two places: `props.isDisabled` and a function name in this component. The function is only used for disabling the dropdown options, not the entire box itself.  So I'd go with renaming the `isDisabled` function to `isOptionDisabled`.

macobo replied:
Agreed, fixed!

RobertJoonas reviewed (commented):
The code looks good to me, but before approving I'd like to verify the UI on staging 👍

macobo commented on assets/js/dashboard/stats/modals/prop-filter-modal.js:
```diff
@@ -1,119 +1,124 @@
-    const {type, clauses} = parseQueryFilter(query, 'props')
+  if (query.filters['props']) {
+    const values = Object.fromEntries(parseQueryFilter(query, 'props').map((value, index) => [index, value]))
+    console.log({ query, values })
```
:facepalm:

RobertJoonas commented on assets/js/dashboard/filters.js:
```diff
@@ -38,19 +38,26 @@ function clearAllFilters(history, query) {
   if (key === "props") {
-    const [[propKey, _propValue]] = Object.entries(query.filters['props'])
-    return <>Property <b>{propKey}</b> {type} {clauses.map(({label}) => <b key={label}>{label}</b>).reduce((prev, curr) => [prev, ' or ', curr])} </>
+    const propTexts = parseQueryFilter(query, key).map(propFilterText)
```
Instead of all prop filters being merged into a single label like
<img width="257" alt="image" src="https://github.com/plausible/analytics/assets/56999674/a2d6fe84-3102-435c-becc-83d30b3b15c7">

What do you think if we'll create separate labels for each `prop_key`?

<img width="334" alt="image" src="https://github.com/plausible/analytics/assets/56999674/07741a5e-f5d5-469d-9c50-ba50be6414a3">

macobo replied:
I agree this is much better!

RobertJoonas reviewed (commented):
Some more comments after having tested the UI out on staging.

* Looks like `contains` filter breaks when multiple clauses are selected.

E.g.:
<img width="396" alt="contains" src="https://github.com/plausible/analytics/assets/56999674/fccc028a-d7ba-47d5-889e-6b2d4004ebdf">

will return 500 in all the reports.

* In the custom prop breakdown report we’re allowing the user to pick any property from the dropdown.

I think only the prop_keys that have hits should be returned. What I mean is, it should not be possible to break down by a prop and then only see the `(none)` value in the returned list.  

I suspect line 137 in filter_suggestions.ex is to blame, since it’s calling `Query.remove_event_filters(query, [:props])`. I don’t think we should do that anymore.

ukutaht commented:
One comment about UX. Take this case:

1. Add a filter for `propA` prop
2. Change prop breakdown key to `propB`
3. Click on one entry from the list

The result is that the filter for the `propA` filter gets overwritten. Instead I think it should keep the `propA` filter and add the `propB` filter as well. Basically from the customers point of view each custom prop is treated like a separate filter. This is also aligned with the visual comments by @RobertJoonas

macobo commented on assets/js/dashboard/stats/modals/prop-filter-row.js:
```diff
@@ -0,0 +1,85 @@
+          values={propKey ? [propKey] : []}
+          onSelect={(value) => onPropKeySelect(id, value)}
+          placeholder={'Property'}
+          disabledOptions={selectedPropKeys}
```
This is great feedback - the 25 limit is especially important not to cause unexpected pain. However unlike regular filter modal, this requires changing the request structure.

This PR is messy as-is and as such I will be solving this in a follow-up PR. Let me know if you disagree!

macobo commented:
PR has been updated @RobertJoonas @ukutaht - please give it another look.

To make reviewing the changes easier here's an overview of new changes:
- I've added support for matches_multiple in https://github.com/plausible/analytics/pull/3719/commits/c517520462a8425de3e3854a9200450c3250b0eb. Not part of the original scope of this PR, but as result of my changes it started throwing 500s instead of silently failing.
- We now show multiple filter boxes for property filters instead of chucking them into a single one. https://github.com/plausible/analytics/pull/3719/commits/5bdc9b8d3b642b6c592f223f8d414054abf8d4b4
- I've fixed some issues around adding/removing filters and related behavior: https://github.com/plausible/analytics/pull/3719/commits/6195a38a8fdbb8170a213843a6591f00962d1e00
- Navigation from behavioral filters rows no longer removes other property filters: https://github.com/plausible/analytics/pull/3719/commits/4ecc344acaedd25b731a2444f279ff8f63c6ff56
- We no longer show properties which only have (none) values in properties behavior tab: https://github.com/plausible/analytics/pull/3719/commits/d38928b0d9ec017bf625363954228236e99f321a

RobertJoonas reviewed (commented):
Nice, we're getting there! Not quite done with feedback yet, sorry... 😄

In addition to the inline comments I had about the code, I've tested things out locally, and noticed a couple of new things:

1. In the prop filter modal, the blue "Add filter" button/link boundaries are huge and it feels a little weird when you click on a blank space and it will append new input fields into the view.
<img width="445" alt="image" src="https://github.com/plausible/analytics/assets/56999674/93b388d7-7c4f-4fbb-91fe-9c27f2730813">

2. It looks like `prop_key` suggestions are returned without the constraint of having them set up in `site.allowed_event_props`. We implemented this to "filter out garbage props" that we don't want displayed there. And another reason is to enforce customers to configure the props they want to see in the dashboard. I made sure it's working on prod, so I suppose it's introduced with this PR?

3. Question: Is it the intended behaviour that the selected property in the behaviours (property breakdown) section will change when you add/remove filters? When I interact with the filters, I don't really expect the breakdown property to change 🤔

https://github.com/plausible/analytics/assets/56999674/f46a28ce-e849-440a-ab44-55e93fefddc9

macobo commented on assets/js/dashboard/filters.js:
```diff
@@ -37,34 +48,38 @@ function clearAllFilters(history, query) {
-    return <>Property <b>{propKey}</b> {type} {clauses.map(({label}) => <b key={label}>{label}</b>).reduce((prev, curr) => [prev, ' or ', curr])} </>
+  if (filterType === "props") {
+    const [{ propKey, clauses, type }] = parseQueryPropsFilter(query).filter((filter) => filter.propKey.value === key)
+    return <>Property <b>{propKey.label}</b> {type} {clauses.map(({label}) => <b key={label}>{label}</b>).reduce((prev, curr) => [prev, ' or ', curr])} </>
```
> This is because custom properties can have any name, and it might collide with the prop names we're using internally.

Note that this code path only deals with _custom_ properties, which we render by its in-code name. No collision with our own properties should be possible here AFAIK.

RobertJoonas replied:
I got the console error locally when I added a custom property `page` and applied the mentioned filters. The below `else if` code path will deal with our own props, and the label for a page filter there is also `page`.

macobo replied:
Thanks! Managed to reproduce, the error was in a slightly different but related spot!

macobo commented on assets/js/dashboard/query.js:
```diff
@@ -65,9 +65,13 @@ export function parseQuery(querystring, site) {
-  return Object.keys(query.filters)
-    .map((key) => [key, query.filters[key]])
-    .filter(([_key, value]) => !!value);
+  const pageKeys = Object.entries(query.filters)
+    .map(([key, value]) => ({ key, value, filterType: key }))
+    .filter(({ key, value }) => key !== 'props' && !!value)
+  const propKeys = Object.entries(query.filters.props || {})
+    .map(([key, value]) => ({ key, value, filterType: 'props' }))
+
+  return pageKeys.concat(propKeys)
```
Good note re: filterType naming clashing with operations. Rather than rename `filterType` to `name` here (which IMO would add confusion) I'll rename FILTER_TYPES to FILTER_OPERATIONS.

macobo commented:
> Nice, we're getting there! Not quite done with feedback yet, sorry... 😄

Thank you for the review, no need to apologize, you are great! :bow: 

> 1. In the prop filter modal, the blue "Add filter" button/link boundaries are huge and it feels a little weird when you click on a blank space and it will append new input fields into the view.

Fixed.

> 2. It looks like prop_key suggestions are returned without the constraint of having them set up in site.allowed_event_props. We implemented this to "filter out garbage props" that we don't want displayed there. And another reason is to enforce customers to configure the props they want to see in the dashboard. I made sure it's working on prod, so I suppose it's introduced with this PR?

That's odd - I didn't change anything here. I tried to reproduce by going to site settings > delete a prop > back to dashboard and the deleted property no longer showed.

> 3. Question: Is it the intended behaviour that the selected property in the behaviours (property breakdown) section will change when you add/remove filters? When I interact with the filters, I don't really expect the breakdown property to change 🤔

Good question. This behavior existed pre-this-pr to avoid showing a blank screen when user would like to break down by a property that's not available. I think this is an easy fix - would you be okay if I fixed this in a separate PR (for easier reviewability)?

RobertJoonas commented:
> That's odd - I didn't change anything here. I tried to reproduce by going to site settings > delete a prop > back to dashboard and the deleted property no longer showed.

Hmm weird, I can't reproduce either anymore. You can ignore it for now then

> Good question. This behavior existed pre-this-pr to avoid showing a blank screen when user would like to break down by a property that's not available. I think this is an easy fix but requires a bit of back-and-forth - would you be okay if I fixed this in a separate PR (for easier reviewability)?

Sure, makes sense!

RobertJoonas reviewed (approved):
Nice! Only docs left 👍

ukutaht commented on assets/js/dashboard/stats/modals/prop-filter-modal.js:
```diff
@@ -1,119 +1,123 @@
+    }
+  } else {
+    return {
+      entries: [0],
```
True, objects are much easier to update immutably in Javascript.

ukutaht reviewed (approved):
🚀
